plutus-chain-index/src/Plutus/ChainIndex/Handlers.hs

```diff
@@ -260,16 +269,49 @@ insertUtxoDb ::
     => UtxoState.UtxoState UtxoState.TxUtxoBalance
     -> Eff effs ()
 insertUtxoDb (UtxoState.UtxoState _ TipAtGenesis) = throwError $ InsertionFailed UtxoState.InsertUtxoNoTip
-insertUtxoDb (UtxoState.UtxoState balance (Tip sl (BlockId bi) (BlockNumber bn)))
-    = addRows (utxoRows db) [UtxoRow (fromIntegral sl) bi bn (toByteString balance)]
+insertUtxoDb (UtxoState.UtxoState (UtxoState.TxUtxoBalance outputs inputs) (Tip sl (BlockId bi) (BlockNumber bn)))
+    = insert $ mempty
+        { tipRows = InsertRows [TipRow (fromIntegral sl) bi bn]
+        , unspentOutputRows = InsertRows $ UnspentOutputRow tipRowId . toByteString <$> Set.toList outputs
+        , unmatchedInputRows = InsertRows $ UnmatchedInputRow tipRowId . toByteString <$> Set.toList inputs
+        }
+        where
+            tipRowId = TipRowId (fromIntegral sl)
 
-deleteOldUtxoDb :: Member DbStoreEffect effs => Tip -> Eff effs ()
-deleteOldUtxoDb TipAtGenesis = pure ()
-deleteOldUtxoDb (Tip slot _ _) = deleteRows $ delete (utxoRows db) (\row -> _utxoRowSlot row <=. val_ (fromIntegral slot))
+reduceOldUtxoDb :: Member DbStoreEffect effs => Tip -> Eff effs ()
+reduceOldUtxoDb TipAtGenesis = pure ()
+reduceOldUtxoDb (Tip slotNo _ _) = do
+    -- Delete all the tips before 'slot'
+    deleteRows $ delete (tipRows db) (\row -> _tipRowSlot row <. val_ slot)
```

Comment:
I would still like to use some kind of a view or some shared value here, but this also works.